docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+   the deprecated function with the `"_v23"` suffix or the one from
+   `Gen.ts`.  If the old one is called, it may have to do some
+   marshalling of request and response (see above).
+
+It will happen that a new client will not be able to accomplish
+something with an old API (example: if video calling is introduced in
+`"v12"`, you can't emulate `POST /video-call` when talking to a `"v9"`
+server.  In these cases, the function in `Man.ts` must raise a "server
+too old" exception, and gracefully shut down the new functionality.
+
+
+## Concerns and design alternatives
+
+### Why not version every end-point separately?
+
+Yes, that would work in principle.  On the backend, it would make the
+entire routing table smaller (no need to concatenate the same
+end-point for many versions), which may result in shorter compile
+times.  On the clients, with a new API version it would be
+straight-forward to see which end-points need to be worked on, and
+which remain unchanged.
+
+On the other hand, the routing table size may not be an issue, and if
+it is there are solutions (introduce a CCP switch to compile only the
+most recent API version that you're working on); and the client
+process is already quite straight-forward with the approach outlined
+above via diffing the swagger docs between most recent version and
+predecessor.
+
+Plus, if the entire API has one version, you get a few advantages:
+
+1. The fact that clients are forced to commit to a concrete API
+   version for all end-points when talking to the backend reduces
+   testing complexity.  If there is a mapping of end-points to
+   versions, the behavior of interacting parties is much less
+   restricted, and versions that have not been tested against each
+   other may be used together.  (This can be avoided, but it's less
+   obvious how to get it right, and testing complexity will likely be
+   worse.)
+
+2. The "one version" approach makes it obvious which end-points are in
+   the most recent API at any given point in time.  The "one version
+   per end-point" approach would either yeild a noisy union of all
+   supported versions, or there would have to be a mechanism for
+   reconstructing something close to what we get for free otherwise.
+
+3. The backend code is a good combination of concise and type-safe in
+   the "one version" approach.  If every end-point had its own
+   version, the routing table entry would either have to accept a
+   variable path segment for the version, and fail at run-time if the
+   version is not supported, or you would have to add one handler per
+   supported version (even if in the case where all versions call the
+   same handler function with slightly different parameters).
+
+
+### Syntactical vs. behavioral changes
+
+It is quite common that behavior of end-points changes together with
+the syntax, or even without a change in the syntax.
+
+This is not a fundamental problem: since the handler can be called
```

Comment:
What happens to old frontend clients if we force a change of version when the behavior changes, but the data doesn't?

E.g. there's an endpoint to upload assets, that are stored in a somehow generic bucket (V1).
A new implementation is introduced so that the image is stored on a per-user bucket (V2).
The endpoint otherwise has the same data between V1 and V2.

Old clients that only talk V1 could, in theory, still use the endpoint to send images. The fact that they are not aware of the new behavior does not prevent them from sending images; both V1 and V2-aware backends and clients can process those images successfully. However, since the client doesn't speak V2, it will be prevented from talking to V2 backends. This sounds to me like an artificial limitation.

Probably in cases like this we should consider changing the behavior of the V1 endpoint to behave like V2, keep the V1, and at the same time introduce V2. Basically backporting the V2 behavior to V1, since they are compatible data-wise. This might increase complexity on the backend implementation (does it?), but has the advantage of not forcing a client update for every user.

Would this be a valid strategy?

Reply:
I would recommend against retroactively changing APIs of old versions. That somewhat defeats the purpose of using versioning, breaks the whole module and can introduce hard-to-diagnose issues.

I don't quite understand your example here. If V2 is the same as V1, but just implemented differently, it should just not cause a version bump, so V1=V2, and clients have no compatibility issues. If V2 has strictly more features than V1, the backend can support both, so V1-clients can still talk to {V1,V2}-backends.

Reply:
thanks @pcapriotti, I agree!  @marcoconti83 let me know if this just raises further questions.  :)

Reply:
My example was a bit convoluted, but you @pcapriotti got what I meant when you said:

>  If V2 has strictly more features than V1, the backend can support both, so V1-clients can still talk to {V1,V2}-backends.

... (adding) thanks to the backend still honoring the V1 API, even if the internals of the backend behavior is changed, if that change is not relevant for the API.

I'll rewrite the example:

With V1, we store assets without any user information (who uploaded them).
The clients use `POST /v1/assets`.

With V2, we introduce a new feature to keep track of which user uploaded which asset, and allow users to delete them. We somehow run a migration script once to assign all previously unassigned assets to a user (doesn't matter how) and from now on, we assume all assets are assigned to a user.
The clients use `POST /v2/assets`, `GET /v2/assets` and `DELETE /v2/assets/<id>`

We now have two options:

- Have the backend support only V2
	- Clean cut, but it also means all clients need to update NOW or they will stop working as they still post to V1
- Have the backend support V1 and V2
	- When clients use `POST /v1/assets`, the backend assigns a user ID to the asset. This was not the behavior before V2 was introduced, BUT the V1 API is still honored (this is what I meant by "backporting" the behavior). The new behavior is not relevant for the API contract, so why deprecate it? We just map `/v1/assets/` to `/v2/assets/`. We can remove the extra v2 bits from the response to look like v1, but that's not even strictly necessary if they were only additions.
	- V2 clients can still do the extra bits: delete assets, even on assets uploaded by V1 clients
	- This means we can release the backend version that supports V2 and wait for all clients to update to V2 whenever they are ready
	
I'm bringing up this example because we had to do such hybrid systems in the past to support a transition between one behavior and another, and I don't want the versioning system to preclude this possibility.

Reply:
Thanks @marcoconti83 for clarifying the example, now I understand better what you mean. I think option two should work without problems. In practice, what I imagine would happen in the server is that V1 would be unmodified at the level of the API and the corresponding types, but its implementation would share code with V2, implementing the new user-tracking functionality.

But I see your point now: in some sense, we are indeed retroactively changing V1 when we add V2, since V1 is now implemented the same as V2. However, this is only in the implementation, and the API *contract* of V1 is still identical to before.

How to make sure that these contracts are indeed preserved over time is another matter altogether, and not discussed in this proposal. We've had some previous discussions around a compatibility testing framework, but it was focused on the server-to-server case, and based on a slightly different approach to versioning: https://wearezeta.atlassian.net/wiki/spaces/CORE/pages/529170506/Compatibility+testing+for+different+backend+versions+in+federation. I imagine many of the ideas still apply, though.